### PR TITLE
Article editor in new open (frontend only)

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
- [x] Testing
  - [x] Code is tested *Lightly: CKEditor is hard to test in our Node-based testing environment since it relies on browser-only APIs*
  - [x] Changes have been manually tested

#### What are the relevant tickets?
#3853 

#### What's this PR do?
This PR:
- Moves all direct usage of CKEditor to a new local package, `ol-ckeditor`
- Changes all CKEditor components exported from `ol-ckeditor` to only use lazy-loading *because CKEditor is quite heavy*.
- Adds a new editor, `CKEditorArticleLazy`, that is similar to the article editor at https://discussions-rc.odl.mit.edu/create_post/; this is currently only used in the frontend and not hooked up to any backend data source.

#### How should this be manually tested?
1. Get `CKEDITOR_ENVIRONMENT_ID`,`CKEDITOR_SECRET_KEY`,`CKEDITOR_UPLOAD_URL` env variables from RC. *These are used for image uploads*.
2. Checkout this branch, `cc/ck-article` and `docker compose up`
3. Visit `/infinite/article` and try out the rich text editor.
    - for now, data is NOT persisted to the backend in any way.
4.  Some things to try:
    - text styles, ordered & un-ordered lists
    - image insertion (dragging OR click the ¶ symbol to left of editor)
    - media embedding ... E.g., youtube videos or any page. Either: (1) click ¶ --> media embed, and enter a URL, or (2) just copy-paste a URL into the editor.

#### Screenshots (if appropriate)

Screenshot 2023-06-21 at 4 26 24 PM

### CODE REVIEW

**Comment by ChristopherChudzicki** [comment]
@Ferdi FYI, see screenshot

File: frontends/infinite-corridor/webpack.config.js [hunk 1]
```diff
@@ -3,10 +3,8 @@ const path = require("path")
 const webpack = require("webpack")
 const BundleTracker = require("webpack-bundle-tracker")
 const MiniCssExtractPlugin = require("mini-css-extract-plugin")
-const CKEditorWebpackPlugin = require("@ckeditor/ckeditor5-dev-webpack-plugin")
 const { BundleAnalyzerPlugin } = require("webpack-bundle-analyzer")
-
-const { styles } = require("@ckeditor/ckeditor5-dev-utils")
+const { withCKEditor } = require("ol-ckeditor/webpack-utils")
```
Comment (ChristopherChudzicki) [inline frontends/infinite-corridor/webpack.config.js:7]: CKEditor requires some special webpack configuration. Almost all of that has been moved to `withCKEditor`.

File: frontends/ol-ckeditor/src/components/CkeditorArticle.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    MediaEmbed,
+    Paragraph,
+    BlockToolbar,
+    UploadAdapter,
+    CloudServices,
+    EasyImage,
+    Image,
+    ImageStyle,
+    ImageToolbar,
+    ImageUpload,
```
Comment (ChristopherChudzicki) [inline frontends/ol-ckeditor/src/components/CkeditorArticle.tsx:51]: You might notice an error in the console:
> index.js:5 Uncaught (in promise) TypeError: Cannot read properties of null (reading 'model')

(A very similar `console.error` appears in OCW Studio.)

As far as I can tell, this has no noticeable consequences other than the console error, and it seems to be a bug in `ckeditor5-react` + ImageUpload plugin; see https://github.com/ckeditor/ckeditor5-react/issues/370

File: frontends/ol-widgets/src/components/editing/ManageWidgetDialog.test.tsx [hunk 3]
```diff
@@ -12,10 +12,10 @@ type TestProps = PartialBy<
   ManageWidgetDialogProps,
   "specs" | "onSubmit" | "onCancel" | "isOpen"
 >
-const setupEditingDialog = (props?: TestProps) => {
+const setupEditingDialog = async (props?: TestProps) => {
```
Comment (ChristopherChudzicki) [inline frontends/ol-widgets/src/components/editing/ManageWidgetDialog.test.tsx:15]: The editor now loads lazily, so some new things get `await`ed here.

File: frontends/ol-widgets/src/index.ts [hunk 4]
```diff
@@ -1,14 +1,2 @@
-/**
- * A TS project consuming this TS source code will try to typecheck this
- * package. Since CKEditor doesn't have great typing, it will fail unless it has
```
Comment (ChristopherChudzicki) [inline frontends/ol-widgets/src/index.ts:3]: CKEditor is now entirely written in Typescript (as of v37) 🎉

File: package.json [hunk 5]
```diff
@@ -78,7 +79,6 @@
         "stylelint-config-standard-scss": "^7.0.1",
         "swc-loader": "^0.2.3",
         "tiny-invariant": "^1.3.1",
-        "ts-jest": "^28.0.5",
```
Comment (ChristopherChudzicki) [inline package.json:81]: We were not actually using this.

File: tsconfig.json [hunk 6]
```diff
@@ -3,7 +3,7 @@
     "incremental": true,
     "composite": true,
     "target": "es6",
-    "module": "es6",
+    "module": "es2020",
```
Comment (ChristopherChudzicki) [inline tsconfig.json:6]: Changed to ES2020 to support dynamic imports (aka lazy loading).

_Changing `module` affects what import/export syntax Typescript will let us use. Webpack still transpiles it to something with wider browser support._

File: frontends/ol-ckeditor/src/components/CkeditorArticle.tsx
Comment (ChristopherChudzicki) [inline frontends/ol-ckeditor/src/components/CkeditorArticle.tsx:1]: This file is basically https://github.com/mitodl/ckeditor-custom-build/blob/master/src/ckeditor.js, updated a bit to account for changes to CKEditor in the last 5 years. (E.g., placeholder text is much easier now.)

**Review by JenniWhitman — APPROVED** [review]
Looks and functions great - The ¶ throws me off for some reason. I know we used that prior, but it seems an odd choice for media insertion - it's the paragraph symbol, no?  Looks like there's a conflict (sorry, comes from my changes), but otherwise, functionally good.

**Comment by ChristopherChudzicki** [comment]
@JenniWhitman re
> Looks and functions great - The ¶ throws me off for some reason. I know we used that prior, but it seems an odd choice for media insertion - it's the paragraph symbol, no? Looks like there's a conflict (sorry, comes from my changes), but otherwise, functionally good.

I agree, and apparently we have an old issue for changing it: https://github.com/mitodl/open-discussions/issues/1632, which I'll assign to me :)

**Pull request merged**